src/Database/LSMTree/Internal/Run/Reader.hs

```diff
@@ -0,0 +1,165 @@
+                     rawBytesToOverflowPages)
+import           Database.LSMTree.Internal.RawPage
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Run.FsPaths
+import           Database.LSMTree.Internal.Serialise
+import qualified System.FS.API as FS
+import           System.FS.API (HasBufFS, HasFS)
+
+-- | Allows reading the k\/ops of a run incrementally, using its own read-only
+-- file handle and in-memory cache of the current disk page.
+--
+-- Creating a 'Reader' does not increase the run's reference count, so make sure
+-- the run remains open while using the 'Reader'.
+--
+-- New pages are loaded when trying to read their first entry.
+--
+-- TODO(optimise): Reuse page buffers using some kind of allocator. However,
+-- deciding how long a page needs to stay around is not trivial.
+data Reader fhandle = Reader {
+      readerCurrentPage    :: !(IORef RawPage)
+      -- | The index of the entry to be returned by the next call to 'next'.
+    , readerCurrentEntryNo :: !(PrimVar RealWorld Word16)
+      -- | Read mode file handle into the run's k\/ops file. We rely on it to
+      -- track the position of the next disk page to read, instead of keeping
+      -- a counter ourselves.
+    , readerKOpsHandle     :: !fhandle
+      -- | The run this reader is reading from.
+    , readerRun            :: !(Run.Run fhandle)
+    }
+
+new ::
+     HasFS IO h
+  -> HasBufFS IO h
+  -> Run.Run (FS.Handle h)
+  -> IO (Reader (FS.Handle h))
+new fs bfs readerRun = do
+    readerKOpsHandle <-
+      FS.hOpen fs (runKOpsPath (Run.lsmRunFsPaths readerRun)) FS.ReadMode
+    readerCurrentEntryNo <- newPrimVar 0
+    -- load first page from disk, if it exists.
+    firstPage <- fromMaybe emptyRawPage <$> readDiskPage fs bfs readerKOpsHandle
+    readerCurrentPage <- newIORef firstPage
+    return Reader {..}
+
+-- | This function should be called when discarding a 'Reader' before it
+-- was done (i.e. returned 'Empty'). This avoids leaking file handles.
+-- Once it has been called, do not use the 'Reader' any more!
+close ::
+     HasFS IO h
+  -> Reader (FS.Handle h)
+  -> IO ()
+close fs Reader {..} = do
+    FS.hClose fs readerKOpsHandle
+
+-- | The 'SerialisedKey' and 'SerialisedValue' point into the in-memory disk
+-- page. Keeping them alive will also prevent garbage collection of the 4k byte
+-- array, so they're long-lived, consider making a copy!
+data Result fhandle
+  = Empty
+  | ReadSmallEntry !SerialisedKey !(Entry SerialisedValue (BlobRef fhandle))
+    -- TODO: Do we need to return all of these fields?
+    -- TODO(optimise): Sometimes, reading the overflow pages is not necessary.
+    -- We could just return the page index and suffix length here to delay doing
+    -- the I\/O.
+  | ReadLargeEntry
+      !SerialisedKey
+      !(Entry SerialisedValue (BlobRef fhandle))
+      !RawPage
+      !Word32
+      ![RawOverflowPage]
```

Comment:
What if `next` had an additional argument that might tell it whether to fully read the pages, or read the blobref into memory already?

Reply:
> > Why are the pages required? I don't really doubt that they are necessary, but could you elaborate with a comment in the code?
> 
> Required as opposed to what? Only returning the prefix and then loading the overflow pages later? That would also work, but it's hard to know upfront whether they will be needed or not (and they usually are), so we just always pass them along for now.

I think what had me confused was that `ReadLargeEntry` has both an `Entry` and a `RawPage`. The `Entry` is the prefix, I think? From the context of this module, that's not really clear, so let's specify that in a comment.

> > What if next had an additional argument that might tell it whether to fully read the pages, or read the blobref into memory already?
> 
> That's complicated, since then the result type depends on an input value. If it makes sense for the caller not to want to load something, I think returning a reference that can then be loaded when needed works best. We could also do that for the overflow pages, and I think I have a `TODO` for it somewhere.

That also works

Reply:
I added a comment to clarify and justify the (redundant) representation.